src/diag.md

```diff
@@ -0,0 +1,177 @@
+# Emitting Diagnostics
+
+A lot of effort has been put into making `rustc` have great error messages.
+This chapter is about how to emit compile errors and lints from the compiler.
+
+## `Span`
+
+`Span` is the primary data structure in `rustc` used to represent a location in
+the code being compiled. `Span`s are attached to most constructs in HIR and MIR,
+allowing for easier error reporting whenever an error comes up.
+
+A `Span` can be looked up in a `CodeMap` to get a "snippet" useful for
+displaying errors with [`span_to_snippet` and other similar methods][sptosnip]
+on the `CodeMap`.
+
+[sptosnip]: https://doc.rust-lang.org/nightly/nightly-rustc/syntax/codemap/struct.CodeMap.html#method.span_to_snippet
+
+## Error messages
+
+The [`rustc_errors`][errors] crate defines most of the utilities used for
+reporting errors.
+
+[errors]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_errors/index.html
+
+Most "session"-like types in the compiler (e.g. [`Session`][session]) have
+methods (or fields with methods) that allow reporting errors. These methods
+usually have names like `span_err` or `struct_span_err` or `span_warn`, etc...
+There are lots of them; they emit different types of "errors", such as
+warnings, errors, fatal errors, suggestions, etc.
+
+[session]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc/session/struct.Session.html
+
+In general, there are two class of such methods: ones that emit an error
+directly and ones that allow finer control over what to emit. For example,
+[`span_err`][spanerr] emits the given error message at the given `Span`, but
+[`struct_span_err`][strspanerr] instead returns a
+[`DiagnosticBuilder`][diagbuild].
+
+`DiagnosticBuilder` allows you to add related notes and suggestions to an error
+before emitting it by calling the [`emit`][emit] method. See the
+[docs][diagbuild] for more info on what you can do.
+
+[spanerr]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc/session/struct.Session.html#method.span_err
+[strspanerr]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc/session/struct.Session.html#method.struct_span_err
+[diagbuild]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_errors/diagnostic_builder/struct.DiagnosticBuilder.html
+[emit]: https://doc.rust-lang.org/nightly/nightly-rustc/rustc_errors/diagnostic_builder/struct.DiagnosticBuilder.html#method.emit
+
+For example, to add a help message to an error, one might do:
+
+```rust,ignore
+let snip = sess.codemap().span_to_snippet(sp);
+
+sess.struct_span_err(sp, "oh no! this is an error!")
+    .span_suggestion(other_sp, "try using a qux here", format!("qux {}", snip))
+    .emit();
+```
+
```

Comment:
Could you also include some stuff on suggestion applicabilities here?

Reply:
I don't want to block too much on this, once this PR is merged it can be expanded more iteratively.